File: Searching Algorithms/Linear Search/Java Implementation/LinearSearchDemo.java
class LinearSearchDemo {
    public static void main(String[] args) {
        int[] numbers = { 12, 23, 43, 6, 67, 24, 12, 90, 1 };

        System.out.println("linearSearchNonRecursive execution");
        // Executing linearSearchNonRecursive method
        displayResult(linearSearchNonRecursive(12, numbers)); // Value found at index 0
        displayResult(linearSearchNonRecursive(24, numbers)); // Value found at index 5
        displayResult(linearSearchNonRecursive(1, numbers)); // Value found at index 8
        displayResult(linearSearchNonRecursive(88, numbers)); // Value not found

        // Testing the validations by passing an empty array and by passing a null array
        displayResult(linearSearchNonRecursive(12, new int[] {})); // Value not found
        displayResult(linearSearchNonRecursive(88, null)); // Value not found

        System.out.println("\nlinearSearchRecursive execution");
        // Executing linearSearchRecursive method
        displayResult(linearSearchRecursive(12, numbers,0)); // Value found at index 0
        displayResult(linearSearchRecursive(24, numbers,0)); // Value found at index 5
        displayResult(linearSearchRecursive(1, numbers,0)); // Value found at index 8
        displayResult(linearSearchRecursive(88, numbers,0)); // Value not found

        // Testing the validations by passing an empty array and by passing a null array
        displayResult(linearSearchRecursive(12, new int[] {},0)); // Value not found
        displayResult(linearSearchRecursive(88, null,0)); // Value not found
    }

    /**
     * evaluates the result index and displays the output. if the index is not
     * equals to -1, then the method displays the output as "Value found at index
     * <index>" (ex: Value found at index 5) or else displays the output "Value not
     * found".
     * 
     * @param resultIndex value to bde resolved (retunred output from linear
     *                    search).
     * @return (Nothing is returned (void) but displays the output on screen).
     */
    public static void displayResult(int resultIndex) {
        System.out.println((resultIndex != -1 ? "Value found at index " + resultIndex : "Value not found"));
    }

    /**
     * This method represents the non recursive implementation of linear search. It
     * searches for the specified key within the given array and returns the result
     * 
     * @param key value which is searched for.
     * @param array array to be searched for the existance of key.
     * @return int index in which the value is located at or -1 if the value is not
     *         available within the array.
     */
    public static int linearSearchNonRecursive(int key, int array[]) {
        // validates the 'array' to identify whether it is not null or has at least one element
        if (array != null && array.length > 0) {

            // loops through the array to search each index value and compares it with key
            for (int i = 0; i < array.length; i++) {
                if (key == array[i]) {
                    return i;
                }
            }
        }
        // returns -1 to indicate the key is not found in any index, if any of the above conditions are not satisfied
        return -1;
    }

    /**
     * This method represents the recursive implementation of linear search. It
     * searches for the specified key within the given array and returns the result.
     * 
     * This one line implementation can be implemented with multi line if else conditions as well
     * 
     * @param key   value which is searched for.
     * @param array array to be searched for the existance of key.
     * @param index parameter which is used to keep track of array indexes while
     *              recursion is in action.
     * @return int index in which the value is located at or -1 if the value is not
     *         available within the array.
     */
    public static int linearSearchRecursive(int key, int array[], int index) {
        return (array != null && array.length > 0 && index < array.length) ? (array[index] == key ? index : linearSearchRecursive(key, array, index += 1)) : -1;
    }
}

File: Searching Algorithms/Linear Search/Python Implementation/linear-search-demo.py
'''
    This method represents the non recursive implementation of linear search. It
    searches for the specified key within the given list and returns the result.

    @param key value which is searched for.
    @param valuesList this is the list to be searched for the availability the of key.
    @return int valuesList index in which the value is located at or -1 if the value is not
             available within the array.
'''
def linear_search_non_recursive(key, valuesList):
    if valuesList is not None and len(valuesList) > 0:
        for index in range(len(valuesList)):
            if key == valuesList[index]:
                return index
    return -1

'''
    This method represents the recursive implementation of linear search. It
    searches for the specified key within the given list and returns the result.

    @param key value which is searched for.
    @param valuesList this is the list to be searched for the availability the of key.
    @param index parameter which is used to keep track of list indexes while recursion is in action.
    @return int valuesList index in which the value is located at or -1 if the value is not
             available within the array.
'''
def linear_search_recursive(key, valuesList, index=0):
    if valuesList is not None and len(valuesList) > 0 and index < len(valuesList):
        return index if valuesList[index] == key else linear_search_recursive(key, valuesList, (index+1))
    return -1

    # Experimental implementation (partially working) for warpping up the above implementation into a single line of code (contribution required)
    # return -1 if valuesList is not None and len(valuesList) > 0 and index < len(valuesList) else (index if valuesList[index] == key else linear_search_recursive(key, valuesList, (index+1)))

'''
    This method evaluvates the output index (which is passed as the resultIndex paramter) from linear search functions and
    if the result is not equals to -1 then a message will be displayed as "Value found at index <index>" (ex : Value found at index 3).

    If the result is -1 then this will produce the output "Value not found"
'''
def display_result(resultIndex):
    print(f'Value found at index {resultIndex}' if (resultIndex is not None and resultIndex != -1) else 'Value not found')


def main():
    numbers = [23, 34, 56, 78, 89, 45, 2, 5]

    print('\nExecuting linear_search_non_recursive function')
    display_result(linear_search_non_recursive(23, numbers)) # Value found at index 0
    display_result(linear_search_non_recursive(78, numbers)) # Value found at index 3
    display_result(linear_search_non_recursive(5, numbers)) # Value found at index 7

    display_result(linear_search_non_recursive(-23, numbers)) # Value not found
    display_result(linear_search_non_recursive(654, numbers)) # Value not found
    display_result(linear_search_non_recursive(322, None)) # Value not found
    display_result(linear_search_non_recursive(433, [])) # Value not found

    print('\nExecuting linear_search_recursive function')
    display_result(linear_search_recursive(23, numbers)) # Value found at index 0
    display_result(linear_search_recursive(78, numbers)) # Value found at index 3
    display_result(linear_search_recursive(5, numbers)) # Value found at index 7

    display_result(linear_search_recursive(-23, numbers)) # Value not found
    display_result(linear_search_recursive(654, numbers)) # Value not found
    display_result(linear_search_recursive(322, None)) # Value not found
    display_result(linear_search_recursive(433, [])) # Value not found
    


if __name__ == '__main__':
    main()
